skbio/sequence/_sequence.py

```diff
@@ -1764,6 +1784,144 @@ def mismatch_frequency(self, other, relative=False):
+            `chars`. Thus, the relative frequencies will not necessarily sum to
+            1.0 if `chars` is provided.
+
+        Returns
+        -------
+        dict
+            Frequencies of characters in the sequence.
+
+        Raises
+        ------
+        TypeError
+            If `chars` is not a ``str`` or ``set`` of ``str``.
+        ValueError
+            If `chars` is not a single-character ``str`` or a ``set`` of
+            single-character strings.
+
+        See Also
+        --------
+        kmer_frequencies
+        iter_kmers
+
+        Notes
+        -----
+        If the sequence is empty (i.e., length zero), ``relative=True``,
+        **and** `chars` is provided, the relative frequency of each specified
+        character will be ``np.nan``.
+
+        If `chars` is not provided, this method is equivalent to, but faster
+        than, ``seq.kmer_frequencies(k=1)``.
+
+        If `chars` is not provided, it is equivalent to, but faster than,
+        passing ``chars=seq.observed_chars``.
+
+        Examples
+        --------
+        Compute character frequencies of a sequence:
+
+        >>> from pprint import pprint
+        >>> from skbio import Sequence
+        >>> seq = Sequence('AGAAGACC')
+        >>> freqs = seq.frequencies()
+        >>> pprint(freqs) # using pprint to display dict in sorted order
+        {'A': 4, 'C': 2, 'G': 2}
+
+        Compute relative character frequencies:
+
+        >>> freqs = seq.frequencies(relative=True)
+        >>> pprint(freqs)
+        {'A': 0.5, 'C': 0.25, 'G': 0.25}
+
+        Compute relative frequencies of characters A, C, and T:
+
+        >>> freqs = seq.frequencies(chars={'A', 'C', 'T'}, relative=True)
+        >>> pprint(freqs)
+        {'A': 0.5, 'C': 0.25, 'T': 0.0}
+
+        Note that since character T is not in the sequence we receive a
+        relative frequency of 0.0. The relative frequencies of A and C are
+        relative to the number of characters in the sequence (8), **not** the
+        number of A and C characters (4 + 2 = 6).
+
+        """
+        # Perform quick sanity checks first before computing frequencies.
+        if chars is not None:
+            chars = self._munge_to_char_set(chars)
+
+        freqs = np.bincount(self._bytes,
+                            minlength=self._number_of_extended_ascii_codes)
+        obs_bytes = np.nonzero(freqs)
+        # Downcast from int64 to uint8 then convert to str. This is safe
+        # because we are guaranteed to have indices in the range 0 to 255
+        # inclusive.
+        obs_chars = obs_bytes[0].astype(np.uint8).tostring().decode('ascii')
+        obs_counts = freqs[obs_bytes]
+        if relative:
+            obs_counts = obs_counts / len(self)
+        freqs = dict(zip(obs_chars, obs_counts))
+
+        if chars is not None:
+            # Get a copy of the keys because we're deleting in the loop.
+            for obs_char in list(freqs.keys()):
```

Comment:
This can probably be vectorized and pushed into the numpy code above. I'll look into this on Monday. Performance-wise there won't be a difference but it'll be cleaner than this.